### PR TITLE
feat!: Range checks and assert in Noir, update to Nargo 0.38, and shamir preprocessing on the fly if needed

### PR DESCRIPTION
**Pull request by rw0x0**

### CODE REVIEW

**Review by dkales — CHANGES_REQUESTED** [review]
The changes in mpc-core are not really breaking, so I would suggest splitting those into a single commit and then doing the remainging things in one commit

File: mpc-core/src/protocols/shamir.rs [hunk 1]
```diff
@@ -287,23 +284,27 @@ impl<F: PrimeField, N: ShamirNetwork> ShamirProtocol<F, N> {
...
-                "not enough correlated random pairs",
-            ))
+        if self.rng_buffer.r_t.is_empty() {
+            debug_assert!(self.rng_buffer.r_2t.is_empty());
+            if self.rng_buffer.num_parties != 3 {
+                // In the 3-party case no communication is required, so we do not print a warning
+                tracing::warn!("Precomputed randomness buffer empty, refilling...");
+            }
+            self.rng_buffer
+                .buffer_triples(&mut self.network, Self::DEFAULT_PAIR_GEN_AMOUNT)?;
```
[the remark this answers is not in the document]
Comment (rw0x0) [reply mpc-core/src/protocols/shamir.rs]: done

**Comment by rw0x0** [comment]
I have created a new branch where I squashed the commits: #270

**Comment by dkales** [comment]
superceded by #270

**Pull request closed without merging**